src/main/java/org/dmfs/jems/generator/elementary/Sequence.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2018 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.generator.elementary;
+
+import org.dmfs.jems.function.Function;
+import org.dmfs.jems.generator.Generator;
+
+
+/**
+ * A sequence {@link Generator}. It generates a sequence by passing the previous result (or initial value) into a given {@link Function}.
+ * <p>
+ * Note, this {@link Generator} always generates the upcoming value ahead of time, which means it generates one more value than you actually need. For very
```

Comment:
Why does it need to generate the upcoming value ahead of time? Is there some benefit you expect from that in some cases?

Reply:
This is just the easiest way to implement this generator.

Reply:
I am not sure I understand this. This is a very simple method in any ways, or do I miss something?
Like:
```java
mNext = mFunction.value(mNext);
return mNext;
```
Shouldn't it be decided based on what we would expect from a type like this?
And since lazy evaluation is used everywhere else in the library, shouldn't this be lazy too?

Reply:
If you implement it like this, you will "skip" the initial value and start with the next one.

Reply:
This implementation is still lazy. I mean, there is no other way to implement in infinite generator other than doing it lazily. It's just slightly less lazy than it could be.

Reply:
Oh, right, silly me.
So it may need a boolean flag then. But it would still be a very low complexity implementation anyways, so I wouldn't think that the expected behavior should be decided based on that few lines.

Reply:
Just reading the note in the javadocs and the little cognitive overhead that that consideration adds each time we use this class is not worth the trade in my opinion between the very simple and extremely simple implementations.

```java
public final class Sequence<T> implements Generator<T>
{
    private T mNext;
    private boolean first = true;
    private final Function<T, T> mFunction;

    public Sequence(T first, Function<T, T> function)
    {
        mNext = first;
        mFunction = function;
    }

    @Override
    public T next()
    {
        if (first)
        {
            first = false;
            return mNext;
        }
        mNext = mFunction.value(mNext);
        return mNext;
    }
}
```

Reply:
How about this:
```java
public final class Sequence<T> implements Generator<T>
{
    private T mNext;
    private Function<T, T> mFunction;

    public Sequence(T first, Function<T, T> function)
    {
        mNext = first;
        mFunction = value -> {
            // the first call returns t as is, afterwards we use the given function
            mFunction = function;
            return value;
        };
    }

    @Override
    public T next()
    {
        mNext = mFunction.value(mNext);
        return mNext;
    }
}
```
Works without a boolean flag, at the expense of a non-final function.

Reply:
What a clever trick! :)
A bit unusual and not so obvious to understand at the first glance, but since it would be well hidden here with nothing else around, and it achieves what's need, it looks fine for me.